Translate "Strict Mode" (#178)
reactjs/ru.reactjs.org: content/docs/strict-mode.md

---
id: strict-mode
title: Строгий режим
permalink: docs/strict-mode.html
---

`StrictMode` – инструмент для обнаружения потенциальных проблем в приложении. Также как и `Fragment`, `StrictMode` не рендерит видимого UI. Строгий режим активирует дополнительные проверки и предупреждения для своих потомков.

> Примечание:
>

> > Проверки строгого режима работают только в режиме разработки; _они не оказывают никакого эффекта в продакшен сборке_.

Reviewer: ```suggestion
> Проверки строгого режима работают только в режиме разработки; _они не оказывают никакого эффекта в продакшен-сборке_.
```

Строгий режим может быть включён для любой части приложения. Например:
`embed:strict-mode/enabling-strict-mode.js`

В примере выше проверки строгого режима *не* будут выполняться для компонентов `Header` и `Footer`. Однако будут выполнены для `ComponentOne` и `ComponentTwo`, а также для всех их потомков.

На данный момент `StrictMode` помогает в:
* [Обнаружении небезопасных методов жизненного цикла](#identifying-unsafe-lifecycles)
* [Предупреждении об использовании устаревшего API строковых реф](#warning-about-legacy-string-ref-api-usage)
* [Предупреждении об использовании устаревшего метода findDOMNode](#warning-about-deprecated-finddomnode-usage)
* [Обраружении неожиданных побочных эффектов](#detecting-unexpected-side-effects)
* [Обнаружении устаревшего API контекста](#detecting-legacy-context-api)

Дополнительные проверки будут включены в будущих релизах React.

### Обнаружение небезопасных методов жизненного цикла {#identifying-unsafe-lifecycles}

> [В данной статье блога](/blog/2018/03/27/update-on-async-rendering.html) рассматриваются причины, почему некоторые методы жизненного цикла небезопасно использовать в асинхронных React приложениях. Если в приложении подключены сторонние библиотеки, то отследить использование таких методов довольно тяжело. К счастью, строгий режим помогает сделать это!

Reviewer: ```suggestion
[В данной статье блога](/blog/2018/03/27/update-on-async-rendering.html) рассматриваются причины, почему некоторые методы жизненного цикла небезопасно использовать в асинхронных React-приложениях. Если в приложении подключены сторонние библиотеки, то отследить использование таких методов довольно тяжело. К счастью, строгий режим помогает сделать это!
```

> [В данной статье блога](/blog/2018/03/27/update-on-async-rendering.html) рассматриваются причины, почему некоторые методы жизненного цикла небезопасно использовать в асинхронных React-приложениях. Если в приложении подключены сторонние библиотеки, то отследить использование таких методов довольно тяжело. К счастью, строгий режим помогает сделать это!

Reviewer 2: ```suggestion
[В этой статье](/blog/2018/03/27/update-on-async-rendering.html) рассматриваются причины, почему некоторые методы жизненного цикла небезопасно использовать в асинхронных React-приложениях. Если в приложении подключены сторонние библиотеки, то отследить использование таких методов довольно тяжело. К счастью, строгий режим помогает сделать это!
```

Когда включён строгий режим, React составляет список всех классовых компонентов, которые используют небезопасные методы жизненного цикла, и отображает информацию о них таким образом:

![](../images/blog/strict-mode-unsafe-lifecycles-warning.png)

Если избавиться от проблем, выявленных в строгом режиме, _уже сегодня_, то это позволит получить все преимущества асинхронного рендеринга в будущих релизах React.

### Предупреждение об использовании устаревшего API строковых реф {#warning-about-legacy-string-ref-api-usage}

> Ранее React предоставлял два способа управления рефами: устаревшие строковые рефы и колбэк API. Хотя строковые рефы и были более удобным способом, они имели [различные недостатки](https://github.com/facebook/react/issues/1373). Поэтому мы рекомендовали [использовать колбэки вместо них](/docs/refs-and-the-dom.html#legacy-api-string-refs).

Reviewer 2: ```suggestion
Ранее React предоставлял два способа управления рефами: устаревшие строковые рефы и колбэк API. Хотя строковые рефы и были более удобным способом, они имели [несколько недостатков](https://github.com/facebook/react/issues/1373). Поэтому мы рекомендовали [использовать колбэки вместо них](/docs/refs-and-the-dom.html#legacy-api-string-refs).
```

> React 16.3 добавил третий способ, удобный как строковые рефы и без каких-либо недостатков:

Reviewer 2: Так на слух может быть чуть лучше.
```suggestion
В React 16.3 добавлен ​​третий способ, который предлагает удобство строковых рефов без каких-либо недостатков:
```

`embed:16-3-release-blog-post/create-ref-example.js`

> Поскольку объекты-рефы были добавлены как замена строковых реф, строгий режим теперь предупреждает об использовании стоковых реф.

Reviewer 2: ```suggestion
Поскольку объекты-рефы стали заменой строковых реф, строгий режим теперь предупреждает об использовании строковых реф.
```

> **Примечание:**
>
> Колбэк-рефы по-прежнему поддерживаются вместе с новым `createRef` API.
>
> Вам не обязательно заменять колбэк-рефы в ваших компонентах. Их использование более гибкое, поэтому они считаются продвинутой возможностью.

[Ознакомьтесь с новым `createRef` API здесь.](/docs/refs-and-the-dom.html)

### Предупреждение об использовании устаревшего метода findDOMNode {#warning-about-deprecated-finddomnode-usage}

Ранее React использовал `findDOMNode` для поиска DOM-узла в дереве по указанному экземпляру класса. В большинстве случаев этот метод не используется, поскольку можно [привязать реф непосредственно к DOM-узлу](/docs/refs-and-the-dom.html#creating-refs).

> `findDOMNode` может использоваться для классовых компонентов, однако это нарушает уровни абстракции, позволяя родительскому компоненту требовать, чтобы происходил рендер определённого дочернего элемента. Это приводит к проблемам при рефакторинге, когда не удаётся изменить детали реализации компонента, так как родитель может использовать DOM-узел этого компонента. `findDOMNode` возвращает только первый дочерний элемент, но с использованием фрагментов компонент может рендерить несколько DOM-узлов. `findDOMNode` выполняет поиск только один раз. Затем метод возвращает ранее полученный результат при вызове. Если дочерний компонент рендерит другой узел, то это изменение никак не отследить. Поэтому `findDOMNode` работает только когда компоненты возвращалют единственный и неизменяемый DOM-узел.

Reviewer: ```suggestion
`findDOMNode` может использоваться для классовых компонентов, однако это нарушает уровни абстракции, позволяя родительскому компоненту требовать, чтобы происходил рендер определённого дочернего элемента. Это приводит к проблемам при рефакторинге, когда не удаётся изменить детали реализации компонента, так как родитель может использовать DOM-узел этого компонента. `findDOMNode` возвращает только первый дочерний элемент, но с использованием фрагментов компонент может рендерить несколько DOM-узлов. `findDOMNode` выполняет поиск только один раз. Затем метод возвращает ранее полученный результат при вызове. Если дочерний компонент рендерит другой узел, то это изменение никак не отследить. Поэтому `findDOMNode` работает, только когда компоненты возвращают единственный и неизменяемый DOM-узел.
```

> `findDOMNode` может использоваться для классовых компонентов, однако это нарушает уровни абстракции, позволяя родительскому компоненту требовать, чтобы происходил рендер определённого дочернего элемента. Это приводит к проблемам при рефакторинге, когда не удаётся изменить детали реализации компонента, так как родитель может использовать DOM-узел этого компонента. `findDOMNode` возвращает только первый дочерний элемент, но с использованием фрагментов компонент может рендерить несколько DOM-узлов. `findDOMNode` выполняет поиск только один раз. Затем метод возвращает ранее полученный результат при вызове. Если дочерний компонент рендерит другой узел, то это изменение никак не отследить. Поэтому `findDOMNode` работает, только когда компоненты возвращалют единственный и неизменяемый DOM-узел.

Reviewer 3: ```suggestion
`findDOMNode` может использоваться для классовых компонентов, однако это нарушает уровни абстракции, позволяя родительскому компоненту требовать, чтобы происходил рендер определённого дочернего элемента. Это приводит к проблемам при рефакторинге, когда не удаётся изменить детали реализации компонента, так как родитель может использовать DOM-узел этого компонента. `findDOMNode` возвращает только первый дочерний элемент, но с использованием фрагментов компонент может рендерить несколько DOM-узлов. `findDOMNode` выполняет поиск только один раз. Затем метод возвращает ранее полученный результат при вызове. Если дочерний компонент рендерит другой узел, то это изменение никак не отследить. Поэтому `findDOMNode` работает, только когда компоненты возвращают единственный и неизменяемый DOM-узел.
```

Reviewer 3: Странно, но эта опечатка уже была испревлена выше...

Author: Пропустил в замечаниях от lex111. Почему-то она не выделена была и заметил только запятую.

Reviewer 3: @ntishkevich а при редактировании нет кнопки подобной "применить предложенные"?

Author: Есть, в тот раз я поправил место, по которому не было замечания, поэтому делал коммит + пуш с машины, а не используя "Commit suggestion"

Reviewer 3: @ntishkevich Понял, спасибо. Я просто еще не до конца понял как это работает и спрашивал с целью уточнить, как лучше предлагать исправления авторам переводов 😄

Reviewer: Лучше по возможности принимать предложения, по-моему так проще.

}
```

> Примечание:
>

> > CSS атрибут [`display: contents`](https://developer.mozilla.org/ru/docs/Web/CSS/display#display_contents) может применяться для исключения узла из раскладки (layout).

Reviewer: ```suggestion
> CSS-выражение [`display: contents`](https://developer.mozilla.org/ru/docs/Web/CSS/display#display_contents) может применяться для исключения узла из раскладки (layout).
```

### Обраружение неожиданных побочных эффектов {#detecting-unexpected-side-effects}

> React работает в две фазы:

Reviewer: ```suggestion
React работает в два этапа:
```

> * **Фаза рендера (render phase)** определяет какие изменения необходимо произвести, например, в DOM. В течение этой фазы React вызывает `render`, а затем сравнивает полученный результат с результатом предыдущего рендера.

Reviewer: ```suggestion
* **Этап рендеринга (render phase)** определяет, какие изменения необходимо произвести, например, в DOM. В течение этого этапа React вызывает `render`, а затем сравнивает полученный результат с результатом предыдущего рендера.
```

> * **Фаза фиксации (commit phase)** – в ней React применяет любые изменения. (В случае React DOM – это фаза, когда React вставляет, обновляет и удаляет DOM-узлы.) В течении этой фазы React вызывает методы жизненного цикла `componentDidMount` и `componentDidUpdate`.

Reviewer: ```suggestion
* **Этап фиксации (commit phase)** – в ней React применяет любые изменения. (В случае React DOM – это этап, когда React вставляет, обновляет и удаляет DOM-узлы.) В течение этого этапа React вызывает методы жизненного цикла `componentDidMount` и `componentDidUpdate`.
```

> Фаза фиксации обычно не занимает много времени, что нельзя сказать про фазу рендера. По этой причине, готовящийся асинхронный режим (который по умолчанию ещё не включён) делит работу на части, периодически останавливает и возобновляет работу во избежании блокировки браузера. Это означает, что в фазе рендера React может вызвать методы жизненного цикла более чем один раз перед фиксацией, либо вызвать их без фиксации (по причине возникновения ошибки или прерывания с большим приоритетом).

Reviewer: ```suggestion
Этап фиксации обычно не занимает много времени, что нельзя сказать про этап рендеринга. По этой причине, готовящийся асинхронный режим (который по умолчанию ещё не включён) делит работу на части, периодически останавливает и возобновляет работу во избежании блокировки браузера. Это означает, что в этапе рендеринга React может вызвать методы жизненного цикла более чем один раз перед фиксацией, либо вызвать их без фиксации (по причине возникновения ошибки или прерывания с большим приоритетом).
```

> Этап фиксации обычно не занимает много времени, что нельзя сказать про этап рендеринга. По этой причине, готовящийся асинхронный режим (который по умолчанию ещё не включён) делит работу на части, периодически останавливает и возобновляет работу во избежании блокировки браузера. Это означает, что в этапе рендеринга React может вызвать методы жизненного цикла более чем один раз перед фиксацией, либо вызвать их без фиксации (по причине возникновения ошибки или прерывания с большим приоритетом).

Reviewer 2: ```suggestion
Этап фиксации обычно не занимает много времени, что нельзя сказать про этап рендеринга. По этой причине, готовящийся асинхронный режим (который по умолчанию ещё не включён) делит работу на части, периодически останавливает и возобновляет работу, чтобы избежать блокировки браузера. Это означает, что на этапе рендеринга React может вызвать методы жизненного цикла более чем один раз перед фиксацией, либо вызвать их без фиксации (из-за возникновения ошибки или прерывания с большим приоритетом).
```

> Фаза рендера включает в себя следующие методы жизненного цикла:

Reviewer: ```suggestion
Этап рендеринга включает в себя следующие методы жизненного цикла:
```

* `constructor`
* `componentWillMount`
* `componentWillReceiveProps`
* `componentWillUpdate`
* `getDerivedStateFromProps`
* `shouldComponentUpdate`
* `render`
* Функции обновления `setState` (первый аргумент)

> Поскольку вышеупомянутые методы могут быть вызваны более одного раза, важно, чтобы они не приводили к каким-либо побочным эффектам. Игнорирование этого правила может привести к множеству проблем, включая утечки памяти и недопустимое состояние приложения. К сожалению, бывает очень трудно обнаружить эти проблемы, поскольку они часто могу быть [недетерминированными](https://ru.wikipedia.org/wiki/%D0%94%D0%B5%D1%82%D0%B5%D1%80%D0%BC%D0%B8%D0%BD%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%BD%D1%8B%D0%B9_%D0%B0%D0%BB%D0%B3%D0%BE%D1%80%D0%B8%D1%82%D0%BC).

Reviewer 2: Может так?
```suggestion
Поскольку вышеупомянутые методы могут быть вызваны более одного раза, важно, чтобы они не приводили к каким-либо побочным эффектам. Игнорирование этого правила может привести к множеству проблем, включая утечки памяти или недопустимое состояние приложения. К сожалению, такие проблемы тяжело обнаружить из-за их [недетерминированности](https://ru.wikipedia.org/wiki/%D0%94%D0%B5%D1%82%D0%B5%D1%80%D0%BC%D0%B8%D0%BD%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%BD%D1%8B%D0%B9_%D0%B0%D0%BB%D0%B3%D0%BE%D1%80%D0%B8%D1%82%D0%BC).
```

Строгий режим не способен автоматически обнаруживать побочные эффекты, но помогает их отследить, сделав более детерминированными. Такое поведение достигается путём двойного вызова следующих методов:

> * метод `constructor` классового компонента

Reviewer: ```suggestion
* Метод `constructor` классового компонента
```

> * метод `render`

Reviewer: ```suggestion
* Метод `render`
```

> * `setState` функции обновления (первый аргумент)

Reviewer: ```suggestion
* Функции обновления `setState` (первый аргумент)
```

Author: Может и на [96 строке](https://github.com/reactjs/ru.reactjs.org/pull/178/files/0b45a42655ae35e6186766c30aae5cd71205f9c8#diff-39c253d8e499ab766bedd67b055cc841R96) изменить заодно?

> * статически метод `getDerivedStateFromProps`

Reviewer: ```suggestion
* Статический метод `getDerivedStateFromProps`
```

> Примечание:
>
> Это применимо только в режиме разработки. _Методы жизненного цикла не вызываются дважды в продакшен-режиме._

Рассмотрим следующий пример:
`embed:strict-mode/side-effects-in-constructor.js`

На первый взгляд данный пример не кажется проблемным. Но если метод `SharedApplicationState.recordEvent` не является [идемпотентным](https://ru.wikipedia.org/wiki/%D0%98%D0%B4%D0%B5%D0%BC%D0%BF%D0%BE%D1%82%D0%B5%D0%BD%D1%82%D0%BD%D0%BE%D1%81%D1%82%D1%8C#%D0%92_%D0%B8%D0%BD%D1%84%D0%BE%D1%80%D0%BC%D0%B0%D1%82%D0%B8%D0%BA%D0%B5), тогда создание этого компонента несколько раз может привести к недопустимому состоянию приложения. Такие труднонаходимые ошибки могут никак не проявить себя во время разработки или быть настолько редкими, что останутся незамеченными.

> Дважды вызывая такие методы как конструктор компонента позволяет строгому режиму легко обнаружить такие проблемы.

Reviewer: ```suggestion
Дважды вызывая такие методы, как конструктор компонента, позволяет строгому режиму легко обнаружить такие проблемы.
```

> Дважды вызывая такие методы, как конструктор компонента, позволяет строгому режиму легко обнаружить такие проблемы.

Reviewer 3: ```suggestion
Двойной вызов таких методов, как конструктор компонента, позволяет строгому режиму легко обнаружить подобные проблемы.
```

Reviewer 3: Как вам такой вариант?
